#91 - Prevent image flicker with stable ids

raghunandankavi2010 opened the pull request:
Using stable ids to prevent image flicker

raghunandankavi2010 committed:
#91 - Prevent image flicker with stable ids 

Using stable ids to prevent image flicker

raghunandankavi2010 commented:
Note:  getItemId returns a hashcode of contents. We could have used message id. But looking at view-source:https://chat.stackoverflow.com/rooms/info/15/android/?tab=stars. You have `<div class="message owner-star" id="message-44171219">`   if you are the star owner and if you are not `<div class="message" id="message-44189234">` making parsing html tags little messy.

Let me know your thoughts on using contents hashcode in getItemId. I am not sure i like the idea myself!.

raghunandankavi2010 commented:
Also would it not be better to use different view types instead of one and hiding views?

Mauker1 commented:
> Also would it not be better to use different view types instead of one and hiding views?

Can you elaborate?

Mauker1 commented:
Also, Raghu, when you make a PR that is supposed to fix an issue, try to use some of the keywords stated [on github docs](https://help.github.com/articles/closing-issues-using-keywords/), such as "fix|close|resolve #issueNumber". This will automatically close the issue once the PR is merged.

raghunandankavi2010 commented:
Thanks mauker. I will use the key words from next time on. 

Say you upload a image it's a image type. So having a separate layout inflated instead of hiding views making it gone or invisible would be better. The same goes for text type messages. Inflating different layouts for different types of messages.

Mauker1 commented:
Gotcha, it sounds like a good idea.

Mauker1 commented:
Regarding using the hashcode instead of the ID, do you think we would also need to change the MessageAdapter as well?

raghunandankavi2010 commented:
Yes. I thought messages id is unique for MessageAdapter. After logging i came to know its not. Even hashcode of content is not enough as two message content can be same. MessageEvent already has a hashcode implementation that we can use. 

If we could parse message id in StarsActivity we will have unique IDs and we could get rid of hash code. We are not parsing IDs in this case and by default they are 0 or -1 so we don't have unique ids in this case. 

I managed to get the message id. So we don't need hashcode now. 

We could extend the different view type inflation to Message Adapter as well.

raghunandankavi2010 commented:
Is there a better way of getting the data in stars screen other than parsing html?.

raghunandankavi2010 committed:
Get message id by parsing the html div class message

Get message id and set that to Event's message id

raghunandankavi2010 committed:
Return message id instead of hashcode of contents.

Return message id instead of hashcode of contents.

raghunandankavi2010 committed:
Remove unsued import statements.

Removed unused import statements

raghunandankavi2010 committed:
Remove return type as it is inferred

Remove return type as it is inferred

TimCastelijns commented on app/src/main/java/com/tristanwiley/chatse/stars/StarsActivity.kt:
```diff
@@ -49,6 +50,7 @@ class StarsActivity : AppCompatActivity() {
             val response = client.newCall(newMessageRequest).execute()
             val doc = Jsoup.parse(response.body().string())
             val messages = doc.getElementsByClass("monologue")
+            Timber.i("messages :%s", messages.size)
```
Use kotlin string interpolation: "messages : ${messages.size}"

TimCastelijns commented on app/src/main/java/com/tristanwiley/chatse/stars/StarsActivity.kt:
```diff
@@ -63,6 +65,18 @@ class StarsActivity : AppCompatActivity() {
+                        if (messageId.isNotEmpty()) {
+                            val id = messageId.substringAfter("message-").toInt()
+                            event.messageId = id
+                            Timber.i("message :%s", id)
```
Same here

TimCastelijns commented on app/src/main/java/com/tristanwiley/chatse/stars/StarsMessageAdapter.kt:
```diff
@@ -46,6 +46,9 @@ import kotlin.collections.ArrayList
+    init{
+        setHasStableIds(true)
+    }
     override fun onBindViewHolder(holder: MessageViewHolder, pos: Int) {
```
Add a blank line between init and the next fun

raghunandankavi2010 committed:
Koltin string interpolation with $ sign

Logging with timer with koltin string interpolation

raghunandankavi2010 committed:
Add space between init block and the next function

Add a space between init block and the next function so its clear to read.

raghunandankavi2010 committed:
Add space between init block and next fun

Add space between init block and the next function in StarsMessageAdapter as well
